src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -557,36 +561,82 @@ private DRSAccessURL signAzureUrl(
   }
 
   private DRSAccessURL signGoogleUrl(
-      SnapshotCacheResult cachedSnapshot, String gsPath, AuthenticatedUserRequest authUser) {
-    Storage storage;
+      SnapshotCacheResult cachedSnapshot,
+      String gsPath,
+      AuthenticatedUserRequest authUser,
+      String userProject) {
+    BlobId locator = GcsUriUtils.parseBlobUri(gsPath);
+
+    BlobInfo blobInfo = BlobInfo.newBuilder(locator).build();
+
+    String signingProject;
+    String signingUser;
+    // If a user specifies a billing project in the request, prefer that over the snapshot's
+    // project.  If a billing project is required to access the data, and it is not provided,
+    // nothing will fail until the user tries to access the signed URL.
+    if (!StringUtils.isEmpty(userProject)) {
+      signingProject = userProject;
+    } else {
+      signingProject = cachedSnapshot.googleProjectId;
+    }
+    if (authUser != null) {
+      signingUser = authUser.getEmail();
+    } else {
+      signingUser = null;
+    }
+
+    Function<Storage, URL> signUrlFunction =
+        storage ->
+            storage.signUrl(
+                blobInfo,
+                URL_TTL.toMinutes(),
+                TimeUnit.MINUTES,
+                getUrlSigningOptions(signingUser, signingProject));
+
+    URL signedUrl;
     if (cachedSnapshot.isSelfHosted) {
```

Comment:
`isSelfHosted` is declared `Boolean` but it can never be `null` (or this code would NPE). To make that clear, I would change `SnapshotCacheResult.isSelfHosted` to be `boolean`. (The same is true of `SnapshotCacheResult.globalFileIds`

Reply:
Yeah, I'm still kinda split brained about primitive types.  I've lost so much time trying to decode NPEs when code tries to assign nulls to primitives rather than on read

Reply:
Using `boolean` makes the code match its intent, which is nice for the reader. It's better for NPEs too since you get it earlier this way, so you're closer to the source of the `null`.